File: distinct-pairs/distinct-pairs.js
function distinctStockPairs(stocksProfit, target) {
    const seen = new Set()
    const distinctPairs = new Set()

    for (let profit of stocksProfit) {
        const complement = target - profit

        if (seen.has(complement)) {
            const pair = [Math.min(profit, complement), Math.max(profit, complement)]
            distinctPairs.add(pair.toString())
        }
        seen.add(profit)
    }
    return distinctPairs.size
}

console.log(distinctStockPairs([5,7,9,13,11,6,6,3,3], 12))

File: distinct-pairs/distinct-pairs.go
package main

import (
	"fmt"
	"strconv"
)

func distinctStockPairs(stocksProfit []int, target int) int {
	seen := make(map[int]bool)
	distinctPairs := make(map[string]bool)

	for _, profit := range stocksProfit {
		complement := target - profit
		if seen[complement] {
			minVal, maxVal := min(profit, complement), max(profit, complement)
			pairKey := strconv.Itoa(minVal) + "," + strconv.Itoa(maxVal)
			distinctPairs[pairKey] = true
		}
		seen[profit] = true
	}

	return len(distinctPairs)
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {
	stocksProfit := []int{5, 7, 9, 13, 11, 6, 6, 3, 3}
	target := 12
	fmt.Println(distinctStockPairs(stocksProfit, target))  // Output: 3
}
